core/src/main/scala/io/finch/EndpointResult.scala

```diff
@@ -60,69 +46,49 @@ sealed abstract class EndpointResult[+A] {
+    case EndpointResult.Matched(_, _, out) =>
+      val promise = Promise[Try[Output[A]]]
+      val io = Effect[F].runAsync(out) {
+        case Right(r) =>
+          IO.pure(promise.setValue(Return(r)))
+        case Left(t) =>
+          IO.pure(promise.setValue(Throw(t)))
+      }
+      io.unsafeRunSync()
+      Some(Await.result(promise, d))
     case _ => None
   }
 
-  /**
-   * Awaits an [[Output]] of the [[Endpoint]] result or throws an exception if an underlying
-   * [[com.twitter.util.Future]] is failed.
-   *
-   * @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(output)` if this endpoint was matched on a given input, `None` otherwise.
-   */
-  final def awaitOutputUnsafe(d: Duration = Duration.Top): Option[Output[A]] =
+  def awaitOutputUnsafe(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[Output[A]] =
     awaitOutput(d).map(toa => toa.get)
 
-  /**
-   * Awaits a value from the [[Output]] wrapped with [[Try]] (indicating if either the
-   * [[com.twitter.util.Future]] is failed or [[Output]] wasn't a payload).
-   *
-   * @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(value)` if this endpoint was matched on a given input, `None` otherwise.
-   */
-  final def awaitValue(d: Duration = Duration.Top): Option[Try[A]] =
+  def awaitValue(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[Try[A]] =
     awaitOutput(d).map(toa => toa.flatMap(oa => Try(oa.value)))
 
-  /**
-   * Awaits a value from the [[Output]] or throws an exception if either an underlying
-   * [[com.twitter.util.Future]] is failed or [[Output]] wasn't a payload.
-   *
-   * @note @note This method is blocking. Never use it in production.
-   *
-   * @return `Some(value)` if this endpoint was matched on a given input,
-   *         `None` otherwise.
-   */
-  final def awaitValueUnsafe(d: Duration = Duration.Top): Option[A] =
+  def awaitValueUnsafe(d: Duration = Duration.Top)(implicit e: Effect[F]): Option[A] =
     awaitOutputUnsafe(d).map(oa => oa.value)
 }
 
 object EndpointResult {
 
-  final case class Matched[A](
+  final case class Matched[F[_], A](
     rem: Input,
     trc: Trace,
-    out: Rerunnable[Output[A]]
-  ) extends EndpointResult[A] {
+    out: F[Output[A]]
+  ) extends EndpointResult[F, A] {
     def isMatched: Boolean = true
   }
 
-  abstract class NotMatched extends EndpointResult[Nothing] {
+  abstract class NotMatched[F[_]] extends EndpointResult[F, Nothing] {
     def isMatched: Boolean = false
   }
 
-  object NotMatched extends NotMatched {
-    final case class MethodNotAllowed(allowed: List[Method]) extends NotMatched
+  object NotMatched extends NotMatched[Pure] {
+    final case class MethodNotAllowed[F[_]](allowed: List[Method]) extends NotMatched[F]
   }
+
+  implicit def covaryEndpointResult[F[_], A](result: EndpointResult[Pure, A]): EndpointResult[F, A] =
```

Comment:
Anything with `Pure` effect (all notmatched) could be converted the same thing with `F` effect, `F` is not supposed to happen anyway. It's used in pattern matching inside of `Endpoint`